src/App.tsx

```diff
@@ -1,61 +1,154 @@
+import { useState, FormEvent } from 'react';
+
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import { TodoList } from './components/TodoList';
+import { User } from './types/User';
+import { Todo } from './types/Todo';
+
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+function getUser(userId: number): User | null {
+  return usersFromServer.find((user) => user.id === userId) || null;
+}
+
+export const todos: Todo[] = todosFromServer.map((todo) => ({
+  ...todo,
+  user: getUser(todo.userId),
+}));
+
+const allowedChars = `0123456789 
+  abcdefghigklmnopqrstuvwxyz
+  абвгдеёжзийклмнопрстуфхцчшщъыьэюяґєії`;
+
+type NewTodo = {
+  title: string;
+  userName: string;
+  isTitleValid: boolean;
+  isUserNameValid: boolean;
+};
+
+const emptyTodo: NewTodo = {
+  title: '',
+  userName: '',
+  isTitleValid: true,
+  isUserNameValid: true,
+};
+
+let nextTodoId = Math.max(...todosFromServer.map((todo) => todo.id)) + 1;
+
+export const App: React.FC = () => {
+  const [todoList, setTodoList] = useState<Todo[]>(todos);
+
+  const [newTodo, setNewTodo] = useState<NewTodo>(emptyTodo);
+
+  const handleChange = (
+    e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>,
+  ) => {
+    const { name, value } = e.target;
+
+    if (
+      name === 'title' &&
+      allowedChars.includes(value.slice(-1).toLowerCase())
+    ) {
+      setNewTodo((todo) => ({
+        ...todo,
+        [name]: value,
+        isTitleValid: true,
+      }));
+    }
+
+    if (name === 'userName') {
+      setNewTodo((todo) => ({
+        ...todo,
+        [name]: value,
+        isUserNameValid: true,
+      }));
+    }
+  };
+
+  const handleAddTodo = (e: FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    if (!newTodo.title || !newTodo.userName) {
+      return setNewTodo((todo) => ({
+        ...todo,
+        isTitleValid: !!newTodo.title,
```

Comment:
Please also check spaces. I entered only spaces and mine todo was added
![image](https://user-images.githubusercontent.com/54936044/220095347-60a1d48b-39ac-42b1-b13b-abd07e44442d.png)

Reply:
Thanks, fixed